File: Week 8/python/2.binary_search_on_answer.py
"""
WEEK 8 - PYTHON DSA
Topic: Binary Search on the Answer
File: 2.binary_search_on_answer.py

CONCEPT:
    Apply binary search not to find a value in an array, but to find the
    minimum/maximum value X such that some monotonic predicate(X) holds.

KEY POINTS:
    - Works whenever predicate(X) is monotonic — once it flips True/False
      it stays that way along the range.
    - Template:
        lo, hi = min_answer, max_answer
        while lo < hi:
            mid = lo + (hi - lo) // 2
            if predicate(mid): hi = mid     # try smaller
            else:              lo = mid + 1 # need larger
        return lo

ALGORITHM / APPROACH:
    Problems demonstrated:
      1. Integer square root: largest x with x*x <= n.
      2. Koko eating bananas: minimum speed so hours <= H.
      3. Ship packages in D days: minimum capacity so days_needed <= D.

PYTHON-SPECIFIC NOTES:
    - math.isqrt(n) is the standard-library integer square root.
    - max(piles) and sum(weights) are convenient bounds.
    - Use ceiling: (pile + speed - 1) // speed.

DRY RUN:
    sqrt(17):
        lo=1 hi=8 mid=5  25>17 hi=4
        lo=1 hi=4 mid=3  9<=17 lo=3
        lo=3 hi=4 mid=4  16<=17 lo=4
        lo=4 hi=4 -> 4

    Koko piles=[3,6,7,11], H=8:
        lo=1 hi=11 mid=6 hours=1+1+2+2=6<=8 hi=6
        lo=1 hi=6 mid=3 hours=1+2+3+4=10>8 lo=4
        lo=4 hi=6 mid=5 hours=1+2+2+3=8<=8 hi=5
        lo=4 hi=5 mid=4 hours=1+2+2+3=8<=8 hi=4
        return 4

COMPLEXITY:
    O(log(range) * cost_of_predicate).
"""

from typing import List


def sqrt_int(n: int) -> int:
    """Largest integer x with x*x <= n."""
    if n < 2:
        return n
    lo, hi = 1, n // 2
    while lo < hi:
        mid = lo + (hi - lo + 1) // 2  # +1 to avoid infinite loop when hi = lo+1
        if mid * mid <= n:
            lo = mid
        else:
            hi = mid - 1
    return lo


def can_finish(piles: List[int], h: int, speed: int) -> bool:
    """Can Koko eat all piles in h hours at given speed?"""
    hours = 0
    for pile in piles:
        hours += (pile + speed - 1) // speed
    return hours <= h


def min_eating_speed(piles: List[int], h: int) -> int:
    lo, hi = 1, max(piles)
    while lo < hi:
        mid = lo + (hi - lo) // 2
        if can_finish(piles, h, mid):
            hi = mid
        else:
            lo = mid + 1
    return lo


def can_ship(weights: List[int], days: int, capacity: int) -> bool:
    current_load = 0
    days_needed = 1
    for w in weights:
        if w > capacity:
            return False
        if current_load + w > capacity:
            days_needed += 1
            current_load = 0
        current_load += w
    return days_needed <= days


def min_ship_capacity(weights: List[int], days: int) -> int:
    lo = max(weights)   # at least the heaviest package
    hi = sum(weights)   # at most ship everything in one day
    while lo < hi:
        mid = lo + (hi - lo) // 2
        if can_ship(weights, days, mid):
            hi = mid
        else:
            lo = mid + 1
    return lo


def main() -> None:
    print("=== Integer Square Root ===")
    for n in [0, 1, 4, 8, 9, 16, 17, 100]:
        print(f"sqrt({n:3d}) = {sqrt_int(n)}")

    print("\n=== Koko Eating Bananas ===")
    piles = [3, 6, 7, 11]
    print(f"Piles: {piles}")
    print(f"Min speed for H=8: {min_eating_speed(piles, 8)}")  # 4

    piles2 = [30, 11, 23, 4, 20]
    print(f"Piles: {piles2}")
    print(f"Min speed for H=5: {min_eating_speed(piles2, 5)}")  # 30
    print(f"Min speed for H=6: {min_eating_speed(piles2, 6)}")  # 23

    print("\n=== Ship Packages ===")
    weights = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    print(f"Weights: {weights}")
    print(f"Min capacity for D=5 days:  {min_ship_capacity(weights, 5)}")   # 15
    print(f"Min capacity for D=10 days: {min_ship_capacity(weights, 10)}")  # 10


if __name__ == "__main__":
    main()


"""
NOTES — Python vs Java:
    - max() and sum() handle Java's Arrays.stream(...).max/.sum verbose forms.
    - math.isqrt(n) is the production-grade integer sqrt; we implement
      manually here to illustrate the technique.
    - Python ints are unbounded — no overflow concern when computing mid.
"""

File: Week 8/cpp/2.binary_search_on_answer.cpp
/*
 * WEEK 8 - C++ DSA
 * Topic: Binary Search on the Answer
 * File: 2.binary_search_on_answer.cpp
 *
 * CONCEPT:
 *     Binary search applied over an answer range rather than an array.
 *     Works whenever a monotonic predicate is available.
 *
 * KEY POINTS:
 *     - Template: lo..hi over plausible answers; while lo<hi: mid=...,
 *       if predicate(mid) hi=mid else lo=mid+1; return lo.
 *     - Examples here: integer sqrt, Koko bananas, ship-packages-in-D-days.
 *
 * ALGORITHM / APPROACH:
 *     See per-problem code; identical template each time.
 *
 * C++-SPECIFIC NOTES:
 *     - Use long long when intermediate products can exceed 2^31.
 *     - std::ranges::max / std::ranges::reduce (C++20+) shorten bounds.
 *     - We rely on <numeric> std::accumulate for the sum bound.
 *
 * DRY RUN:
 *     sqrt(17): converges to 4 (since 4*4=16<=17<25=5*5).
 *     Koko piles=[3,6,7,11] H=8 -> speed 4.
 *     Ship weights 1..10 D=5 -> capacity 15.
 *
 * COMPLEXITY:
 *     O(log(range) * predicate cost).
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>

int sqrtInt(int n) {
    if (n < 2) return n;
    int lo = 1, hi = n / 2;
    while (lo < hi) {
        int mid = lo + (hi - lo + 1) / 2;
        long long sq = static_cast<long long>(mid) * mid;
        if (sq <= n) lo = mid;
        else         hi = mid - 1;
    }
    return lo;
}

bool canFinish(const std::vector<int>& piles, int h, int speed) {
    long long hours = 0;
    for (int p : piles) hours += (p + speed - 1) / speed;
    return hours <= h;
}

int minEatingSpeed(const std::vector<int>& piles, int h) {
    int lo = 1;
    int hi = *std::max_element(piles.begin(), piles.end());
    while (lo < hi) {
        int mid = lo + (hi - lo) / 2;
        if (canFinish(piles, h, mid)) hi = mid;
        else                          lo = mid + 1;
    }
    return lo;
}

bool canShip(const std::vector<int>& weights, int days, int capacity) {
    int currentLoad = 0, daysNeeded = 1;
    for (int w : weights) {
        if (w > capacity) return false;
        if (currentLoad + w > capacity) {
            ++daysNeeded;
            currentLoad = 0;
        }
        currentLoad += w;
    }
    return daysNeeded <= days;
}

int minShipCapacity(const std::vector<int>& weights, int days) {
    int lo = *std::max_element(weights.begin(), weights.end());
    int hi = std::accumulate(weights.begin(), weights.end(), 0);
    while (lo < hi) {
        int mid = lo + (hi - lo) / 2;
        if (canShip(weights, days, mid)) hi = mid;
        else                              lo = mid + 1;
    }
    return lo;
}

int main() {
    std::cout << "=== Integer Square Root ===\n";
    for (int n : {0, 1, 4, 8, 9, 16, 17, 100}) {
        std::cout << "sqrt(" << n << ") = " << sqrtInt(n) << "\n";
    }

    std::cout << "\n=== Koko Eating Bananas ===\n";
    std::vector<int> piles = {3, 6, 7, 11};
    std::cout << "Piles: [3,6,7,11]\n";
    std::cout << "Min speed for H=8: " << minEatingSpeed(piles, 8) << "\n";  // 4

    std::vector<int> piles2 = {30, 11, 23, 4, 20};
    std::cout << "Piles: [30,11,23,4,20]\n";
    std::cout << "Min speed for H=5: " << minEatingSpeed(piles2, 5) << "\n"; // 30
    std::cout << "Min speed for H=6: " << minEatingSpeed(piles2, 6) << "\n"; // 23

    std::cout << "\n=== Ship Packages ===\n";
    std::vector<int> weights = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    std::cout << "Weights: [1..10]\n";
    std::cout << "Min capacity for D=5 days:  " << minShipCapacity(weights, 5) << "\n";  // 15
    std::cout << "Min capacity for D=10 days: " << minShipCapacity(weights, 10) << "\n"; // 10

    return 0;
}

/*
 * NOTES — C++ vs Java:
 *     - std::max_element returns an iterator; deref with *.
 *     - long long defends against integer overflow in mid*mid.
 *     - std::accumulate is the STL sum.
 */
